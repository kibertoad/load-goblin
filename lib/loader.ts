import { Dirent } from 'fs'
import { isMatch } from 'matcher'

import { promises as fs } from 'fs'
import { resolve } from 'path'
import { safeLoad } from 'js-yaml'

export type Filter = (file: string) => boolean

export type Config = {
  rootDirectory: string
  recursively?: boolean
  filter?: string | Filter
}

export type LoadingResult = {
  content: any
  name: string
  path: string
}[]

async function getFilesInDirectory(directory: string) {
  const allEntries: Dirent[] = await fs.readdir(directory, { withFileTypes: true })
  const files: string[] = []
  const directories: string[] = []

  allEntries.forEach((entry) => {
    if (entry.isFile()) {
      files.push(entry.name)
    }
    if (entry.isDirectory()) {
      directories.push(entry.name)
    }
  })

  return {
    files,
    directories,
  }
}

async function loadContent(path: string): Promise<any> {
  if (path.toLowerCase().endsWith('json')) {
    return require(path)
  }

  const fileContents = await fs.readFile(path, 'utf8')
  if (path.toLowerCase().endsWith('yml') || path.toLowerCase().endsWith('yaml')) {
    return safeLoad(fileContents)
  }

  return fileContents
}

function normalizePath(originalPath: string): string {
  return originalPath.replace(/\\/g, '/')
}

async function accumulateFiles(
  result: LoadingResult,
  currentDirectory: string,
  pathSoFar: string,
  filter: Filter,
  recursively = false
): Promise<void> {
  const resolvedDirectoryPath = resolve(pathSoFar, currentDirectory)
  const directoryEntries = await getFilesInDirectory(resolvedDirectoryPath)

  for await (const file of directoryEntries.files) {
    const resolvedFullPath = resolve(resolvedDirectoryPath, file)

    if (filter(resolvedFullPath)) {
      const content = await loadContent(resolvedFullPath)
      result.push({
        path: normalizePath(resolvedDirectoryPath),
        name: file,
        content,
      })
    }
  }

  if (recursively) {
    for await (const directory of directoryEntries.directories) {
      await accumulateFiles(result, directory, resolvedDirectoryPath, filter, recursively)
    }
  }
}

function resolveFilter(config: Config): Filter {
  if (config.filter instanceof Function) {
    return config.filter as Filter
  }

  if (typeof config.filter === 'string') {
    const normalizedPattern: string = normalizePath(config.filter)

    return (file: string): boolean => {
      return isMatch(normalizePath(file), normalizedPattern)
    }
  }

  return () => true
}

export async function loadFiles(config: Config): Promise<LoadingResult> {
  const result: LoadingResult = []

  const filter = resolveFilter(config)
  await accumulateFiles(result, config.rootDirectory, '', filter, config.recursively)

  return result
}
