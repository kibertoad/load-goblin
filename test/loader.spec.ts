import { loadFiles } from '../lib/loader'
import { resolve } from 'path'

describe('loader', () => {
  function getPath(rootDir = '') {
    return resolve(__dirname, 'files', rootDir)
  }

  it('loads files non-recursively', async () => {
    const filesLoaded = await loadFiles({ rootDirectory: getPath() })
    expect(filesLoaded).toMatchSnapshot()
  })

  it('loads files recursively', async () => {
    const filesLoaded = await loadFiles({ rootDirectory: getPath(), recursively: true })
    expect(filesLoaded).toMatchSnapshot()
  })

  it('loads files recursively with text filter', async () => {
    const filesLoaded = await loadFiles({
      rootDirectory: getPath(),
      recursively: true,
      filter: '*/abc*.json',
    })
    expect(filesLoaded).toMatchSnapshot()
  })

  it('loads files recursively with text filter (ignores slash style)', async () => {
    const filesLoaded = await loadFiles({
      rootDirectory: getPath(),
      recursively: true,
      filter: '*\\abc*.json',
    })
    expect(filesLoaded).toMatchSnapshot()
  })

  it('loads files recursively with function filter', async () => {
    const filesLoaded = await loadFiles({
      rootDirectory: getPath(),
      recursively: true,
      filter: (file) => {
        return file.includes('def')
      },
    })
    expect(filesLoaded).toMatchSnapshot()
  })
})
