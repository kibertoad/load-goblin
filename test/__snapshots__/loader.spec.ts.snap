// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader loads files non-recursively 1`] = `
Array [
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.json",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": "data=content
",
    "name": "root.txt",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.yaml",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.yml",
    "path": "C:/sources/node/load-goblin/test/files",
  },
]
`;

exports[`loader loads files recursively 1`] = `
Array [
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.json",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": "data=content
",
    "name": "root.txt",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.yaml",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "root.yml",
    "path": "C:/sources/node/load-goblin/test/files",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc.json",
    "path": "C:/sources/node/load-goblin/test/files/abc",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc1.json",
    "path": "C:/sources/node/load-goblin/test/files/abc/abc1",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "def.json",
    "path": "C:/sources/node/load-goblin/test/files/def",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "def1.json",
    "path": "C:/sources/node/load-goblin/test/files/def/def1",
  },
]
`;

exports[`loader loads files recursively with function filter 1`] = `
Array [
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "def.json",
    "path": "C:/sources/node/load-goblin/test/files/def",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "def1.json",
    "path": "C:/sources/node/load-goblin/test/files/def/def1",
  },
]
`;

exports[`loader loads files recursively with text filter (ignores slash style) 1`] = `
Array [
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc.json",
    "path": "C:/sources/node/load-goblin/test/files/abc",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc1.json",
    "path": "C:/sources/node/load-goblin/test/files/abc/abc1",
  },
]
`;

exports[`loader loads files recursively with text filter 1`] = `
Array [
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc.json",
    "path": "C:/sources/node/load-goblin/test/files/abc",
  },
  Object {
    "content": Object {
      "data": "content",
    },
    "name": "abc1.json",
    "path": "C:/sources/node/load-goblin/test/files/abc/abc1",
  },
]
`;
